File: Chapter- 10/digital_tree.py
class Trie_Node:

    def __init__(self):
        self.children = [None] * 26

        self.is_EndOfword = False


class Trie:

    def __init__(self):
        self.root = self.getNode()

    def getNode(self):

        return Trie_Node()

    def char_To_Index(self, ch):


        return ord(ch) - ord('a')

    def insert_key(self, key):


        pCrawl = self.root
        length = len(key)
        for level in range(length):
            index = self.char_To_Index(key[level])


            if not pCrawl.children[index]:
                pCrawl.children[index] = self.getNode()
            pCrawl = pCrawl.children[index]

        pCrawl.is_EndOfword = True

    def search_key(self, key):


        pCrawl = self.root
        length = len(key)
        for level in range(length):
            index = self.char_To_Index(key[level])
            if not pCrawl.children[index]:
                return False
            pCrawl = pCrawl.children[index]

        return pCrawl.is_EndOfword


def main():

    keys = ["hi", "this", "is", "heyy", "for",
            "bye"]
    output = ["Not present in trie", "Present in trie"]

    tr = Trie()

    for key in keys:
        tr.insert_key(key)

    print("{} ---- {}".format("hi", output[tr.search_key("hi")]))
    print("{} ---- {}".format("this", output[tr.search_key("this")]))
    print("{} ---- {}".format("is", output[tr.search_key("is")]))
    print("{} ---- {}".format("bye", output[tr.search_key("bye")]))


if __name__ == '__main__':
    main()

File: Chapter- 10/digital_tree.cpp
#include <bits/stdc++.h>
using namespace std;

const int AlphabetSize = 26;
struct trie_Node
{
	struct trie_Node *children[AlphabetSize];

	bool isEndOfWord;
};

struct trie_Node *getNode(void)
{
	struct trie_Node *pNode = new trie_Node;

	pNode->isEndOfWord = false;

	for (int i = 0; i < AlphabetSize; i++)
		pNode->children[i] = NULL;

	return pNode;
}

void insert(struct trie_Node *root, string key)
{
	struct trie_Node *pCrawl = root;

	for (int i = 0; i < key.length(); i++)
	{
		int index = key[i] - 'a';
		if (!pCrawl->children[index])
			pCrawl->children[index] = getNode();

		pCrawl = pCrawl->children[index];
	}

	pCrawl->isEndOfWord = true;
}

bool search(struct trie_Node *root, string key)
{
	struct trie_Node *pCrawl = root;

	for (int i = 0; i < key.length(); i++)
	{
		int index = key[i] - 'a';
		if (!pCrawl->children[index])
			return false;

		pCrawl = pCrawl->children[index];
	}

	return (pCrawl->isEndOfWord);
}

int main()
{
	string keys[] = {"hi", "this", "that",
					"happy", "any", "but",
					"bye", "new" };
	int num = sizeof(keys)/sizeof(keys[0]);

	struct trie_Node *root = getNode();

	for (int i = 0; i < num; i++)
		insert(root, keys[i]);

	search(root, "hi")? cout << "Yes\n" :
						cout << "No\n";
	search(root, "this")? cout << "Yes\n" :
						cout << "No\n";
	search(root, "that")? cout << "Yes\n" :
						cout << "No\n";
	search(root, "happy")? cout << "Yes\n" :
						cout << "No\n";
	search(root, "think")? cout << "Yes\n" :
						cout << "No\n";
	return 0;
}
